// @nex3/market-management/src/models/game-config.js
const mongoose = require("mongoose");
const modelName = "game_configs";

/**
 * @type {Model.GameConfigsModel}
 */
const Schema = new mongoose.Schema({
  name: { type: String, required: true },
  enabled: { type: Boolean, required: true },
  minbet: { type: Number, required: true },
  maxbet: { type: Number, required: true },
  disabledround: { type: Number, required: false },
  botenabled: { type: Boolean, default: false },
  maxbot: { type: Number, required: true },
  bot_maxbet: { type: Number, required: false },
  bot_minbet: { type: Number, required: false },
  hour_maxbot: { type: Array, required: false },
  disabled_message: { type: String, required: false },
  choices_maxbet: { type: Object, required: false },
});

Schema.index({ name: 1 }, { unique: true, background: true });

/**
 * @type {Model.GameConfigsRepository}
 */
const repository = mongoose.models[modelName] ?? mongoose.model(modelName, Schema);

repository.getByName = (name) => repository.findOne({ name }, "-_id -__v -time").lean().exec();

module.exports = repository;
