// @nex3/market-management/src/util.js
const template = "@nex3/market-controller";

const cacheKeys = {
  allMarkets: `${template}.market.all`,
  gameMarket: (gameId) => `${template}.market.${gameId}`,
  allChips: `${template}.chip.all`,
  gameConfigs: (gameId) => `${template}.gameConfigs.${gameId}`,
  allGameMarkets: `${template}.gameMarket.all`,
};

const gameConfigsKeyMapping = {
  name: "name",
  enabled: "enabled",
  minbet: "minBet",
  maxbet: "maxBet",
  disabledround: "disableRound",
  botenabled: "botEnabled",
  maxbot: "maxBot",
  bot_maxbet: "maxBetForBot",
  bot_minbet: "minBetForBot",
  hour_maxbot: "percentOfBotPerHour",
  disabled_message: "disabledMessage",
  choices_maxbet: "maxBetChoices",
};

const normalizeGameConfigs = (gameConfigs) =>
  Object.assign(
    {},
    ...Object.keys(gameConfigs).map((key) => {
      const newKey = gameConfigsKeyMapping[key] || key;

      return { [newKey]: gameConfigs[key] };
    })
  );

const normalizeChip = (chip) => {
  delete chip._id;
  delete chip.__v;
  delete chip.time;
  delete chip.enabled;

  return chip;
};

module.exports = { cacheKeys, normalizeGameConfigs, normalizeChip };
